rust_src/src/lists.rs

```diff
@@ -406,6 +373,47 @@ impl LispCons {
     }
 }
 
+impl LispStructuralEqual for LispCons {
+    fn equal(
+        &self,
+        other: Self,
+        kind: equal_kind::Type,
+        depth: i32,
+        ht: &mut LispHashTableRef,
+    ) -> bool {
+        let (circular_checks, item_depth) = if kind == equal_kind::EQUAL_NO_QUIT {
+            (LispConsCircularChecks::off, 0)
+        } else {
+            (LispConsCircularChecks::on, depth + 1)
+        };
+
+        // This is essentially a zip. However, one cons can end earlier than the other.
+        // Which requires that the iterators are available to call `rest()` on.
+        // Had `.zip()` been used this would not be possible as it consumes them.
+        let mut it1 = self.iter_tails(LispConsEndChecks::off, circular_checks);
+        let mut it2 = other.iter_tails(LispConsEndChecks::off, circular_checks);
+        loop {
+            match (it1.next(), it2.next()) {
+                (Some(cons1), Some(cons2)) => {
+                    let (item1, tail1) = cons1.into();
+                    let (item2, tail2) = cons2.into();
+                    if item1.equal_internal(item2, kind, item_depth, ht) {
+                        if tail1.eq(tail2) {
+                            return true;
+                        }
+                    } else {
+                        return false;
+                    }
+                }
+                (None, None) => break,
+                _ => return false,
+            }
+        }
+
+        it1.rest().equal_internal(it2.rest(), kind, depth + 1, ht)
```

Comment:
Won't `it1.rest()` and `it2.rest()` both be empty at this point?

Reply:
Not if the list was an "improper" list of `(a . b)` style.

Reply:
That is why the whole thing is done with a `loop` and not some other construct. The iterators have to be left intact to examine and other methods would have consumed them. I should add a comment. Thanks.

Reply:
Ah, indeed. Yes, a comment about what might be left over would be nice.